src/Symfony/Component/Serializer/Serialize/DataModel/DataModelBuilder.php

```diff
@@ -0,0 +1,112 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\Serialize\DataModel;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Component\Serializer\Exception\LogicException;
+use Symfony\Component\Serializer\Exception\MaxDepthException;
+use Symfony\Component\Serializer\Php\ArgumentsNode;
+use Symfony\Component\Serializer\Php\FunctionCallNode;
+use Symfony\Component\Serializer\Php\MethodCallNode;
+use Symfony\Component\Serializer\Php\PhpNodeInterface;
+use Symfony\Component\Serializer\Php\PropertyNode;
+use Symfony\Component\Serializer\Php\ScalarNode as PhpScalarNode;
+use Symfony\Component\Serializer\Php\VariableNode;
+use Symfony\Component\Serializer\Serialize\Config\SerializeConfig;
+use Symfony\Component\Serializer\Serialize\Mapping\PropertyMetadataLoaderInterface;
+use Symfony\Component\Serializer\Serialize\VariableNameScoperTrait;
+use Symfony\Component\Serializer\Type\Type;
+use Symfony\Component\VarExporter\ProxyHelper;
+
+/**
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ *
+ * @experimental in 7.0
+ */
+final class DataModelBuilder implements DataModelBuilderInterface
+{
+    use VariableNameScoperTrait;
+
+    public function __construct(
+        private readonly PropertyMetadataLoaderInterface $propertyMetadataLoader,
+        private readonly ContainerInterface $runtimeServices,
+    ) {
+    }
+
+    public function build(Type $type, PhpNodeInterface $accessor, SerializeConfig $config, array $context = []): DataModelNodeInterface
+    {
+        if ($type->isObject() && $type->hasClass()) {
+            $className = $type->className();
+
+            $context['depth_counters'][$className] ??= 0;
+            ++$context['depth_counters'][$className];
+
+            if ($context['depth_counters'][$className] > $config->maxDepth()) {
+                throw new MaxDepthException($className, $config->maxDepth());
```

Comment:
noone is catching this right?

Reply:
No indeed, I do think that the exception must be handled in the user land, to prevent to hide unexpected infinite relationship for example.